test/integration/agent/openflow_test.go

```diff
@@ -422,6 +425,183 @@ func TestNetworkPolicyFlows(t *testing.T) {
+	}
+}
+
+func prepareProxyServiceGroupAndFlows(gid uint32, svc svcConfig, endpointList []k8sproxy.Endpoint, stickyAge uint16) (tableFlows []expectTableFlows, groupBuckets []string) {
+	nw_proto := 6
+	learnProtoField := "NXM_OF_TCP_DST[]"
+	if svc.protocol == ofconfig.ProtocolUDP {
+		nw_proto = 17
+		learnProtoField = "NXM_OF_UDP_DST[]"
+	} else if svc.protocol == ofconfig.ProtocolSCTP {
+		nw_proto = 132
+		learnProtoField = "OXM_OF_SCTP_DST[]"
+	}
+	cookieAllocator := cookie.NewAllocator(roundInfo.RoundNum)
+	svcFlows := expectTableFlows{tableID: 41, flows: []*ofTestUtils.ExpectFlow{
+		{
+			MatchStr: fmt.Sprintf("priority=200,%s,reg4=0x10000/0x70000,nw_dst=%s,tp_dst=%d", string(svc.protocol), svc.ip.String(), svc.port),
+			ActStr:   fmt.Sprintf("group:%d", gid),
+		},
+		{
+			MatchStr: fmt.Sprintf("priority=190,%s,reg4=0x30000/0x70000,nw_dst=%s,tp_dst=%d", string(svc.protocol), svc.ip.String(), svc.port),
+			ActStr:   fmt.Sprintf("learn(table=40,idle_timeout=%d,priority=200,delete_learned,cookie=0x%x,eth_type=0x800,nw_proto=%d,%s,NXM_OF_IP_DST[],NXM_OF_IP_SRC[],load:NXM_NX_REG3[]->NXM_NX_REG3[],load:NXM_NX_REG4[0..15]->NXM_NX_REG4[0..15],load:0x2->NXM_NX_REG4[16..18],load:0x1->NXM_NX_REG0[19]),load:0x2->NXM_NX_REG4[16..18],goto_table:42", stickyAge, cookieAllocator.RequestWithObjectID(4, gid).Raw(), nw_proto, learnProtoField),
+		},
+	}}
+	epDNATFlows := expectTableFlows{tableID: 42, flows: []*ofTestUtils.ExpectFlow{}}
+	hairpinFlows := expectTableFlows{tableID: 106, flows: []*ofTestUtils.ExpectFlow{}}
+	groupBuckets = make([]string, 0)
+	for _, ep := range endpointList {
+		epIP := ipToHexString(net.ParseIP(ep.IP()))
+		epPort, _ := ep.Port()
+		bucket := fmt.Sprintf("weight:100,actions=load:%s->NXM_NX_REG3[],load:0x%x->NXM_NX_REG4[0..15],load:0x2->NXM_NX_REG4[16..18],load:0x1->NXM_NX_REG0[19],resubmit(,42)", epIP, epPort)
+		groupBuckets = append(groupBuckets, bucket)
+
+		unionVal := (0b010 << 16) + uint32(epPort)
+		epDNATFlows.flows = append(epDNATFlows.flows, &ofTestUtils.ExpectFlow{
+			MatchStr: fmt.Sprintf("priority=200,%s,reg3=%s,reg4=0x%x/0x7ffff", string(svc.protocol), epIP, unionVal),
+			ActStr:   fmt.Sprintf("ct(commit,table=50,zone=65520,nat(dst=%s:%d),exec(load:0x21->NXM_NX_CT_MARK[])", ep.IP(), epPort),
+		})
+
+		if ep.GetIsLocal() {
+			hairpinFlows.flows = append(hairpinFlows.flows, &ofTestUtils.ExpectFlow{
+				MatchStr: fmt.Sprintf("priority=200,ip,nw_src=%s,nw_dst=%s", ep.IP(), ep.IP()),
+				ActStr:   "set_field:169.254.169.252->ip_src,load:0x1->NXM_NX_REG0[18],goto_table:110",
+			})
+		}
+	}
+
+	tableFlows = []expectTableFlows{svcFlows, epDNATFlows, hairpinFlows}
+	return
+}
+
+func ipToHexString(ip net.IP) string {
+	ipStr := "0x"
+	ipBytes := ip
+	if ip.To4() != nil {
+		ipBytes = []byte(ip)[12:16]
+	}
+	for idx, ipByte := range ipBytes {
+		if ipByte < 16 && idx > 0 {
+			ipStr = fmt.Sprintf("%s0%x", ipStr, ipByte)
+		} else {
+			ipStr = fmt.Sprintf("%s%x", ipStr, ipByte)
+		}
+	}
+	return ipStr
```

Comment:
I think `return "0x" + hex.EncodeToString(ipBytes)` will work?

https://play.golang.org/p/zQ79mNuNKgi

Reply:
Thanks for the suggestion, updated.